### ISSUE TITLE
Add support for hosted services in .NET MAUI apps

### ISSUE THREAD

**Issue by Eilon** [dotnet/maui#2244]
Look into using [hosted services](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihostedservice?view=dotnet-plat-ext-5.0). This would replace the icky `IMauiInitializeService` that was added in https://github.com/dotnet/maui/pull/2137, and also to enable users to use hosted services. MAUI will automatically call StartAsync/StopAsync at the appropriate time. Sadly, this pattern is async-only, whereas in MAUI the call sites are always (or at least almost always) sync. But given that this is app startup/shutdown, doing async-over-sync might be OK (we would block).

**Comment by pjmagee** [comment]
Just to clarify my understanding, this would mean once supported i could drop the following icky code I'm using?

```
 /*
             * Hosted services only works with the Hosting Extensions from .NET
             * Maui does not use the same Hosting service and so we much register them differently depending on how we execute background services.
             * If this is for Maui, we must hook into Application startup and run the services.
             * If this is for Console, .AddHostedService will handle starting the services for us.
             */
            if (environment.ApplicationName.Equals("HeroesProfile.Console"))
            {
                services
                    .AddHostedService<GameSimulator>()
                    .AddHostedService<FileWatchers>()
                    .AddHostedService<OnLaunchReplayProcessor>();
            }
            else
            {
                services
                    .AddSingleton<GameSimulator>()
                    .AddSingleton<FileWatchers>()
                    .AddSingleton<OnLaunchReplayProcessor>();
            }
```

**Comment by Eilon** [comment]
@pjmagee yeah I think so. If we implement this, there would be proper support for hosted services so all the regular APIs that exist for that would work as expected (starting/stopping those services).

**Comment by RayKoopa** [comment]
Dumb question hopefully not too off-topic; what does an `IHostedService` mean on Android? Will that behave like an _Android foreground service_ or will it only live as long as the MAUI app is running? I want to run an always-on .NET 5 `BackgroundService` on my Android smartphone and am a little lost figuring out how to utilize MAUI or net6.0-android for that yet.

**Comment by Eilon** [comment]
@RayKoopa - This would be for in-app in-proc services. So they would start when the app starts, and dies when the app dies. I'm not sure in Android terminology what that equates to, but it's not a background thing.

**Comment by RayKoopa** [comment]
@Eilon Ah alright, I expected it to be "app-wide". I'll try to figure out how to host my Android foreground service then (which is like a "system-wide" background thread with a notification, that has no connection to an app or UI once started). Thanks for the fast reply!

**Comment by hflexgrig** [comment]
Any news from this issue?

**Comment by hflexgrig** [comment]
@Eilon Added pull request for this issue  https://github.com/dotnet/maui/pull/3718

**Comment by Eilon** [comment]
@hflexgrig oh wow I'll take a look!

**Comment by hflexgrig** [comment]
@Eilon deleted old one and create new https://github.com/dotnet/maui/pull/3724

**Comment by obenyaala** [comment]
any status on this feature? Thanks

**Comment by ctigrisht** [comment]
will there ever be hosted services?

**Comment by kevinchalet** [comment]
FWIW, having native `IHostedService` support would allow OpenIddict users to use the new `OpenIddict.Client.Windows` integration on MAUI/WinUI. I posted more information here: https://github.com/openiddict/openiddict-core/pull/1480#issuecomment-1397746558.

**Comment by JohnHunterACS** [comment]
This kind of addition would definitely make it easier to work with an abstracted service model instead of platform specific service implementations.

**Comment by DoubleDBE** [comment]
Any news about this? Currently stuck because of this...

**Comment by Eilon** [comment]
This work is not currently planned. Could you manually start the services in `MauiProgram` or another part of the app? Certainly not ideal, but it could be a reasonable workaround.

**Comment by janseris** [comment]
> This work is not currently planned. Could you manually start the services in `MauiProgram` or another part of the app? Certainly not ideal, but it could be a reasonable workaround.

Does this really work in Android when screen is switched off?

**Comment by bdovaz** [comment]
Any updates?

**Comment by symbiogenesis** [comment]
Need this

**Comment by wearemeeters** [comment]
I'm also looking forward to seeing it in MAUI. In the meantime, what is the best practice for MAUI to have background tasks that periodically perform REST requests to get data?

**Comment by mariuszdybala** [comment]
I'm also looking forward to seeing it in MAUI. In the meantime, what is the best practice for MAUI to have background tasks that periodically perform REST requests to get data?

**Comment by pjmagee** [comment]
> I'm also looking forward to seeing it in MAUI. In the meantime, what is the best practice for MAUI to have background tasks that periodically perform REST requests to get data?

I think you can take the concept of a hosted service but you need to activate those in the Maui Startup event. that's basically the best workaround we have, not sure the impact on different devices though.

**Comment by Haukinger** [comment]
Hosted services will run while the app is active in the foreground only, right?

**Comment by symbiogenesis** [comment]
> I'm also looking forward to seeing it in MAUI. In the meantime, what is the best practice for MAUI to have background tasks that periodically perform REST requests to get data?

I use a singleton service that initializes a dispatcher timer on a certain interval, which is wired up in the constructor of the service.

A hosted service would be better, for a number of reasons. But that is adequate-ish.

**Comment by illuminat3** [comment]
Any update?

**Comment by MacKey-255** [comment]
Any update?

**Comment by jakubmaguza** [comment]
You can grab [this file](https://github.com/dotnet/aspnetcore/blob/3ea008c80d5cc63de7f90ddfd6823b7b006251ff/src/Hosting/Hosting/src/Internal/HostedServiceExecutor.cs) and register it as HostedService:

`builder.Services.AddSingleton<HostedServiceExecutor>();`
Then just before returning app, you can call it

```
using (var scope = app.Services.CreateScope())
{
    var hsExecutor = scope.ServiceProvider.GetRequiredService<HostedServiceExecutor>();
        hsExecutor.StartAsync(default).GetAwaiter().GetResult();
}
return app;
```
After that it start all your Hosted/Background services.

**Comment by ThaDaVos** [comment]
Is there any update on this? I thought this would be possible as MAUI used a hosted service like structure
